Gems/MotionMatching/Code/Source/FeatureMatrixMinMaxScaler.cpp

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <AzCore/std/string/conversions.h>
+#include <AzFramework/IO/LocalFileIO.h>
+#include <Allocators.h>
+#include <limits>
+#include <FeatureMatrixMinMaxScaler.h>
+
+namespace EMotionFX::MotionMatching
+{
+    AZ_CLASS_ALLOCATOR_IMPL(MinMaxScaler, MotionMatchAllocator, 0)
+
+    bool MinMaxScaler::Fit(const FeatureMatrix& featureMatrix, const Settings& settings)
+    {
+        const FeatureMatrix::Index numRows = featureMatrix.rows();
+        const FeatureMatrix::Index numColumns = featureMatrix.cols();
+
+        m_clip = settings.m_clip;
+        m_featureMin = settings.m_featureMin;
+        m_featureMax = settings.m_featureMax;
+        m_featureRange = settings.m_featureMax - settings.m_featureMin;
+        AZ_Assert(m_featureRange > s_epsilon, "Feature range too small. This will lead to divisions by zero.");
+        m_dataMin.clear();
+        m_dataMin.resize(numColumns, std::numeric_limits<float>::max());
+        m_dataMax.clear();
+        m_dataMax.resize(numColumns, -std::numeric_limits<float>::max());
+
+        for (FeatureMatrix::Index row = 0; row < numRows; ++row)
+        {
+            for (FeatureMatrix::Index column = 0; column < numColumns; ++column)
+            {
+                m_dataMin[column] = AZStd::min(m_dataMin[column], featureMatrix(row, column));
+                m_dataMax[column] = AZStd::max(m_dataMax[column], featureMatrix(row, column));
+            }
+        }
+
+        m_dataRange.clear();
+        m_dataRange.resize(numColumns);
+        for (FeatureMatrix::Index column = 0; column < numColumns; ++column)
+        {
+            const float columnMin = m_dataMin[column];
+            const float columnMax = m_dataMax[column];
+            const float range = columnMax - columnMin;
+            m_dataRange[column] = range;
+        }
```

Comment:
As a peformance optimization, instead of clearing out and resizing the m_dataRange container with value initialized floats which would all be 0.0f, you could instead use the resize_no_construct extension to set the new size of the DataRange container and leave the old data there as the for loop is going to overwrite it anyway.

```suggestion
        m_dataRange.resize_no_construct(numColumns);
        for (FeatureMatrix::Index column = 0; column < numColumns; ++column)
        {
            const float columnMin = m_dataMin[column];
            const float columnMax = m_dataMax[column];
            const float range = columnMax - columnMin;
            m_dataRange[column] = range;
        }
```

Reply:
Good call, changed